pkg/controller/control.go

```diff
@@ -1244,29 +1247,137 @@ func (c *Controller) handleErrorNoLock(err error) error {
 	return err
 }
 
-func (c *Controller) handleDiskNoSpaceErrorForReplicas(replicaNoSpaceErrMap map[string]string) error {
+func (c *Controller) areAllReplicasNoSpace(replicaNoSpaceErrMap map[string]int) bool {
 	if len(replicaNoSpaceErrMap) == 0 {
-		return nil
+		return false
 	}
 
-	replicaWritableCount := 0
-	for _, r := range c.replicas {
-		if r.Mode == types.ERR {
-			continue
-		}
-		replicaWritableCount++
+	rwReplicaCount, rwReplicasOnEnospcMap, noSpaceWOReplicasList := c.categorizeOutOfSpaceReplicas(replicaNoSpaceErrMap)
+	for _, address := range noSpaceWOReplicasList {
+		c.setReplicaModeNoLock(address, types.ERR)
 	}
 
 	// https://github.com/longhorn/longhorn/issues/10718#issuecomment-3190666485
 	// If a replica is on a disk that is out of space, and others are not, this replica will be marked as ERR to avoid the inconsistent data.
-	// The volume will be readable when all disks of the replicas are out of space so last one replica will not be marked as ERR because of ENOSPC.
-	if len(replicaNoSpaceErrMap) == replicaWritableCount {
-		return types.ErrNoSpaceLeftOnDevice
+	// Last one replica will not be marked as ERR for an ENOSPC error so the volume will still be readable when all disks of the replicas are out of space.
+	// If all replicas are on disks that are out of space at the same time, we need to ensure the data integrity by written bytes that each replica has written when encountering ENOSPC
+	// And we can rebuild the less replicas efficiently when ENOSPC issue is fixed.
+	// Therefore, it will retain a list of max number of replicas that have the same write bytes RW and other replicas will be set to ERR.
+	if len(rwReplicasOnEnospcMap) == rwReplicaCount {
```

Comment:
Why can `len(rwReplicasOnEnospcMap)` be not equal to `rwReplicaCount`?

Reply:
3 RW replicas, 2 encounter the ENOSPC error.
- `rwReplicaCount` is 3.
- `len(rwReplicasOnEnospcMap)` is 2.

Reply:
Can you elaborate it in the comment? Thank you.